### ISSUE TITLE
bee doesn't properly handle @some_function() calls

### ISSUE THREAD

**Issue by bugfolder** [backdrop-contrib/b#341]
I discovered this when using `bee cc all` on a local site with CiviCRM (cue ominous music) and PHP 8.1.13, which threw this set of errors:

> mysite % bee cc all
>filemtime(): stat failed for /var/www/ousabd/webroot/modules/contrib/civicrm/vendor/composer/installed.json
 /mysite/modules/contrib/civicrm/vendor/symfony/config/Resource/FileResource.php:62
>
>Cannot modify header information - headers already sent by (output started at /myuser/Sites/bee/bee.php:67)
 /mysite/core/includes/bootstrap.inc:1637
>
>Error: Class "Civi\FlexMailer\Services" not found in flexmailer_civicrm_container() (line 50 of /mysite/modules/contrib/civicrm/ext/flexmailer/flexmailer.php).

However, clearing all caches from the UI didn't create the same errors. The first error is thrown from this line of code:

```
        return false !== ($filemtime = @filemtime($this->resource)) && $filemtime <= $timestamp;
```

Note that `@filemtime` is `@`ed, so it shouldn't be throwing an error when the resource is missing.

A simpler way to reproduce the error is to enable the Devel module, then use the "Execute PHP" command to execute this command:

```
echo 'Test: ' . @filemtime('foobar');
```

That will print "Test" with no error, as expected. However, the `bee` equivalent command,

```
bee ev "echo 'Test: ' . @filemtime('foobar');"
```

emits the error.

This is because `bee` has its own error handler, `bee_error_handler()`, which currently does anticipate some problems of this nature:

```
function bee_error_handler($errno, $message, $filename, $line, array $context = NULL) {
  if (error_reporting() > 0) {
    // Core uses the @ error operator in url_stat() to suppress the warning for
    // non-existent files. However, since PHP 8.0, certain errors are no longer
    // suppressed which causes unnecessary error messages to appear in some
    // 'bee' commands.
    if (version_compare(PHP_VERSION, '8', '>=')) {
      $trace = debug_backtrace();
      if (isset($trace[2]) && $trace[2]['function'] == 'url_stat') {
        return;
      }
    }
    echo "$message\n";
    echo " $filename:$line\n\n";
  }
}
```

However, it only considers the core function `url_stat()`. We need it to also consider `filemtime()` (and maybe more?) to work with CiviCRM.

But there's even a problem with this the way it is. Insert `echo var_dump($trace);` just after the `debug_backtrace()`, re-run the problematic `bee` command, and we get this:

```
array(5) {
  [0]=>
  array(2) {
    ["function"]=>
    string(17) "bee_error_handler"
    ["args"]=>
    array(4) {
      [0]=>
      int(2)
      [1]=>
      string(35) "filemtime(): stat failed for foobar"
      [2]=>
      string(64) "/myuser/Sites/bee/commands/php.bee.inc(51) : eval()'d code"
      [3]=>
      int(1)
    }
  }
  [1]=>
  array(4) {
    ["file"]=>
    string(64) "/myuser/Sites/bee/commands/php.bee.inc(51) : eval()'d code"
    ["line"]=>
    int(1)
    ["function"]=>
    string(9) "filemtime"
    ["args"]=>
    array(1) {
      [0]=>
      string(6) "foobar"
    }
  }
  [2]=>
  array(3) {
    ["file"]=>
    string(44) "/myuser/Sites/bee/commands/php.bee.inc"
    ["line"]=>
    int(51)
    ["function"]=>
    string(4) "eval"
  }
  [3]=>
  array(4) {
    ["file"]=>
    string(44) "/myuser/Sites/bee/includes/command.inc"
    ["line"]=>
    int(91)
    ["function"]=>
    string(17) "eval_bee_callback"
    ["args"]=>
    array(2) {
      [0]=>
      array(1) {
        ["code"]=>
        string(37) "echo 'Test: ' . @filemtime('foobar');"
      }
      [1]=>
      array(0) {
      }
    }
  }
  [4]=>
  array(4) {
    ["file"]=>
    string(31) "/myuser/Sites/bee/bee.php"
    ["line"]=>
    int(30)
    ["function"]=>
    string(19) "bee_process_command"
    ["args"]=>
    array(0) {
    }
  }
}
filemtime(): stat failed for foobar
 /myuser/Sites/bee/commands/php.bee.inc(51) : eval()'d code:1
```

Note that the name of the `@`ed function is in `$trace[1]`, not `$trace[2]`. So I don't think this would work even in its original intended use.

So I suggest two changes:

* First, use `$trace[1]`.
* Second, create an array of functions for which this check needs to be made, and use `in_array()` to do the check. Then, as new functions are discovered, they can be added to the array.

A PR will follow shortly.

**Comment by bugfolder** [comment]
PR submitted; but it's not ideal. It does suppress that warning, but does so even if the function isn't quieted (i.e., even if the @ is missing). So leaving this as "needs work", and I'll work on it.

As a simple check, 

```bee ev "echo filemtime('foobar');"``` 

should issue a warning, but 

```bee ev "echo @filemtime('foobar');"``` 

should not.

**Comment by yorkshire-pudding** [comment]
Thanks for reporting this @bugfolder - I don't have anything to suggest here so will let you look further at this and wait your updated PR.

**Comment by bugfolder** [comment]
I took a look at how Backdrop handles quieted functions, and it does so by checking the PHP `error_reporting()` value. So, rather than checking on a function-by-function basis, we can just use the same comparison that `_backdrop_error_handler_real()` uses.

But then, that led to the thought, "why not just pattern `bee`'s error handler exactly after Backdrop's?" so that they behave as similar as possible. So that's what I've done in the latest PR. Basically copying the call chain from `_backdrop_error_handler()` and converting to `bee` equivalents.

And then this seems to work for quieted functions. Here's how it works on my test site:

Non-quieted:

```
myuser@mysite % bee ev "filemtime('foobar');"

 [!] Warning: filemtime(): stat failed for foobar
in eval() (line 1 of /myuser/Sites/bee/commands/php.bee.inc(51) : eval()'d code). 
```

Quieted:

```
myuser@mysite % bee ev "@filemtime('foobar');"
myuser@mysite % 
```

I did one thing that I wasn't sure of: I used `bee_message()` for the output (so that I could use the different message types), but I saw that the original error handler just used `echo` for its output, and so wondered if I should stick to that in case there are reportable errors that don't get to the point of `bee_print_messages()`.

Feedback on that, or any other aspect of the implementation, is solicited.

**Comment by yorkshire-pudding** [comment]
Perhaps `bee_render_text()` would be best here.  pure `echo` sometimes causes issues and generates "headers already sent" messages.

I have been thinking about whether `bee_render_text()` should also support icons as there's a few things in `dl` that would work better showing as they happen rather than at the end.

I haven't yet tested this, but I think this approach makes sense as long as it doesn't break anything.

**Comment by bugfolder** [comment]
I see that `bee_render_text()` calls `echo bee_format_text()`, and so I used the latter to more streamline the output in the latest update to the PR.

In order to use the icons/colors of `bee_render_text()`, I abstracted the symbols and colors into their own function `bee_message_type_info()`, which makes them accessible to other functions (like the error handler). Hope that's OK. I used an unkeyed array, but I could imagine making it a keyed array, using the same keys as are used later in the processing.
